### PR TITLE
add godep manifest files to staging repos

### PR DESCRIPTION
**Pull request by deads2k**
The staging repos should have manifests that match the godeps of kube so we know what they build against.  We don't need the actual vendored code, since a sync script on the other side needs to find the correct level of other staging directories and thus requires its own `godep restore && go get && godep save` cycle.

@sttts ptal
@lavalamp @caesarxuchao client-go needs a lot of unwinding to do something similar, but the idea is that you can run an acyclic path to get this updated by copying the types and dependencies with `go list`, then generate the clients, then generate this manifest.  Then in your sync script you can pull the proper levels and finish the actual vendoring.

### CODE REVIEW

**Comment by caesarxuchao** [comment]
> since a sync script on the other side needs to find the correct level of other staging directories and thus requires its own godep restore && go get && godep save cycle

FYI this godep cycle is only run for client-go currently. apimachinery doesn't run this cycle.

**Comment by caesarxuchao** [comment]
Agree that staging folders need their own godep manifests.

client-go currently doesn't need to be included in hack/update-staging-godeps.sh because it's running its own `godep restore/save` in its copy.sh.

File: hack/verify-staging-imports.sh [hunk 1]
```diff
@@ -41,10 +41,23 @@ function print_forbidden_imports () {
...
-		echo
-		echo "${FORBIDDEN}" | sed 's/^/  /'
-		echo
-		return 1
+        echo
+        echo "${FORBIDDEN}" | sed 's/^/  /'
+        echo
+        return 1
+    fi
+    local FORBIDDEN=$(
```
Comment (sttts) [inline hack/verify-staging-imports.sh]: please add a comment or at least rename it to TEST_FORBIDDEN

File: hack/verify-staging-imports.sh [hunk 2]
```diff
@@ -41,10 +41,23 @@ function print_forbidden_imports () {
...
+        go list -f $'{{with $package := .ImportPath}}{{range $.TestImports}}{{$package}} imports {{.}}\n{{end}}{{end}}' ./vendor/k8s.io/${PACKAGE}/... |
+        sed 's|^k8s.io/kubernetes/vendor/||;s| k8s.io/kubernetes/vendor/| |' |
+        grep -v " k8s.io/${PACKAGE}" |
+        grep -e "imports \(${RE}\)"
+    )
+    if [ -n "${FORBIDDEN}" ]; then
+        echo "${PACKAGE} has a forbidden dependency:"
```
Comment (sttts) [inline hack/verify-staging-imports.sh]: also here add something about that it's due to tests

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest_test.go [hunk 3]
```diff
@@ -36,9 +36,11 @@ import (
...
 	"k8s.io/apiserver/pkg/endpoints/request"
 	"k8s.io/apiserver/pkg/registry/rest"
-	"k8s.io/kubernetes/pkg/api"
-	"k8s.io/kubernetes/pkg/api/testapi"
-	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/client-go/pkg/api"
+	"k8s.io/client-go/pkg/api/v1"
+
+	// need to register pods
+	_ "k8s.io/client-go/pkg/api/install"
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest_test.go]: how much I dislike these global registrations....

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver_test.go [hunk 4]
```diff
@@ -88,13 +91,14 @@ func setUp(t *testing.T) (*etcdtesting.EtcdTestServer, Config, *assert.Assertion
...
 	config.LoopbackClientConfig = &restclient.Config{}
 
-	config.OpenAPIConfig = DefaultOpenAPIConfig(openapigen.GetOpenAPIDefinitions, api.Scheme)
-	config.OpenAPIConfig.Info = &spec.Info{
-		InfoProps: spec.InfoProps{
-			Title:   "Kubernetes",
-			Version: "unversioned",
-		},
-	}
+	// TODO restore this test, but right now, eliminate our cycle
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver_test.go]: any issue opened for that?

File: hack/update-staging-godeps.sh [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+SKIP_RESTORE=${SKIP_RESTORE:-}
+if [ "${SKIP_RESTORE}" != "true" ]; then
+	echo "starting godep restore"
+	mkdir -p "${godepRestoreDir}"
+
+	# add the vendor folder so that we don't redownload things during restore
+	GOPATH="${godepRestoreDir}:${KUBE_ROOT}/staging:${ORIGINAL_GOPATH}"
+	# restore from kubernetes godeps to ensure we get the correct levels
+	# you get errors about the staging repos not using a known version control system
+	godep restore > ${godepRestoreDir}/godep-restore.log	2>&1 || true
```
Comment (sttts) [inline hack/update-staging-godeps.sh]: nit: space, no tab

File: hack/update-staging-godeps.sh [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+	echo "finished godep restore"
+fi
+
+echo "forcing fake godep folders to match the current state of master in tmp"
+for stagingRepo in $(ls ${KUBE_ROOT}/staging/src/k8s.io); do
+	echo "    creating ${stagingRepo}"
+	rm -rf ${godepRestoreDir}/src/k8s.io/${stagingRepo}
+	cp -R ${KUBE_ROOT}/staging/src/k8s.io/${stagingRepo} ${godepRestoreDir}/src/k8s.io
```
Comment (sttts) [inline hack/update-staging-godeps.sh]: `cp -a`

File: hack/update-staging-godeps.sh [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+	# we need a git commit in the godep folder, otherwise godep won't save
+	pushd ${godepRestoreDir}/src/k8s.io/${stagingRepo}
+	git init > /dev/null
+	# we need this so later commands work, but nothing should ever actually include these commits
+	git config user.email "you@example.com"
```
Comment (sttts) [inline hack/update-staging-godeps.sh]: :scream:

Comment (sttts) [inline hack/update-staging-godeps.sh]: Just pass them to git: `git -c user.email=... -c user.name=... commit -m ...`

Comment (sttts) [inline hack/update-staging-godeps.sh]: Ok, the configs are not global I guess. So maybe it's less 😱  than I thought first.

File: hack/update-staging-godeps.sh [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+	git config user.name "Your Name"
+	git add . > /dev/null
+	git commit -m "fake commit" > /dev/null
```
Comment (sttts) [inline hack/update-staging-godeps.sh]: `-q`

File: hack/update-staging-godeps.sh [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+	popd
+done
+
+function updateGodepManifest() {
+	repo=${1}
```
Comment (sttts) [inline hack/update-staging-godeps.sh]: convention: `local repo=...`

File: hack/update-staging-godeps.sh [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+
+	echo "starting ${repo} save"
+	mkdir -p ${TARGET_DIR}/src/k8s.io
+	# if target_dir isn't the same as source dir, you need copy
+	cp -R ${KUBE_ROOT}/staging/src/${repo} ${TARGET_DIR}/src/k8s.io || true
```
Comment (sttts) [inline hack/update-staging-godeps.sh]: `|| true` swallows errors. not nice

Comment (deads2k) [inline hack/update-staging-godeps.sh]: > || true swallows errors. not nice

Sometimes source and target are the same.  If you give me an easy conditional to detect when they are different, I'll use that instead.

File: hack/verify-staging-godeps.sh [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+# Register function to be called on EXIT to remove folder.
+function cleanup {
+	SKIP_CLEANUP=${SKIP_CLEANUP:-}
+	if [ "${SKIP_RESTORE}" != "true" ]; then
+		rm -rf "${TARGET_DIR}"
+	fi
+}
+trap cleanup EXIT
+
+TARGET_DIR=${TARGET_DIR} ${KUBE_ROOT}/hack/update-staging-godeps.sh
```
Comment (sttts) [inline hack/verify-staging-godeps.sh]: I have never seen this style to export a variable for the sub-process. But seems to work.

File: hack/verify-staging-godeps.sh [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+# check each staging repo to make sure its godeps.json is correct
```
Comment (sttts) [inline hack/verify-staging-godeps.sh]: nit: Godeps.json

File: hack/verify-staging-godeps.sh [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+for stagingRepo in $(ls ${KUBE_ROOT}/staging/src/k8s.io); do
+	# we have to skip client-go because it does unusual manipulation of its godeps
+	if [ "${stagingRepo}" == "client-go" ]; then
+		continue
+	fi
+
+	if ! diff -q ${KUBE_ROOT}/staging/src/k8s.io/${stagingRepo}/Godeps/Godeps.json ${TARGET_DIR}/src/k8s.io/${stagingRepo}/Godeps/Godeps.json; then
+		diff ${KUBE_ROOT}/staging/src/k8s.io/${stagingRepo}/Godeps/Godeps.json ${TARGET_DIR}/src/k8s.io/${stagingRepo}/Godeps/Godeps.json
```
Comment (sttts) [inline hack/verify-staging-godeps.sh]: isn't diff silent if there are no changes? Just do: `! diff ...` without the if-then

File: staging/godeps-json-updater.go [hunk 14]
```diff
@@ -68,6 +68,7 @@ func main() {
 	// removes the Deps whose ImportPath contains "k8s.io/kubernetes"
 	i := 0
 	for _, dep := range g.Deps {
+		// TODO make this a parameter
```
Comment (sttts) [inline staging/godeps-json-updater.go]: please 👍

File: hack/update-staging-godeps.sh (hunk 10 above)
Comment (sttts) [inline hack/update-staging-godeps.sh]: `test "${KUBE_ROOT}/staging" = "${TARGET_DIR}" || cp -a ....`

Or use `rsync -a`. rsync does not care.

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest_test.go (hunk 3 above)
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest_test.go]: > how much I dislike these global registrations....

Agree, but this stops the bleeding.

File: staging/src/k8s.io/apiserver/pkg/server/genericapiserver_test.go (hunk 4 above)
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/genericapiserver_test.go]: > any issue opened for that?

https://github.com/kubernetes/kubernetes/issues/41748

File: staging/godeps-json-updater.go (hunk 14 above)
Comment (deads2k) [inline staging/godeps-json-updater.go]: > please

opened https://github.com/kubernetes/kubernetes/issues/41749

**Comment by deads2k** [comment]
comments addressed.

File: hack/update-staging-godeps.sh [hunk 15]
```diff
@@ -46,34 +46,35 @@ if [ "${SKIP_RESTORE}" != "true" ]; then
...
+	godep restore > ${godepRestoreDir}/godep-restore.log 2>&1 || true
 	echo "finished godep restore"
 fi
 
 echo "forcing fake godep folders to match the current state of master in tmp"
 for stagingRepo in $(ls ${KUBE_ROOT}/staging/src/k8s.io); do
 	echo "    creating ${stagingRepo}"
 	rm -rf ${godepRestoreDir}/src/k8s.io/${stagingRepo}
-	cp -R ${KUBE_ROOT}/staging/src/k8s.io/${stagingRepo} ${godepRestoreDir}/src/k8s.io
+	cp -Ra ${KUBE_ROOT}/staging/src/k8s.io/${stagingRepo} ${godepRestoreDir}/src/k8s.io
```
Comment (sttts) [inline hack/update-staging-godeps.sh]: `-a` includes `-R`

File: hack/update-staging-godeps.sh [hunk 16]
```diff
@@ -46,34 +46,35 @@ if [ "${SKIP_RESTORE}" != "true" ]; then
 	GOPATH="${godepRestoreDir}:${KUBE_ROOT}/staging:${ORIGINAL_GOPATH}"
 	# restore from kubernetes godeps to ensure we get the correct levels
 	# you get errors about the staging repos not using a known version control system
-	godep restore > ${godepRestoreDir}/godep-restore.log	2>&1 || true
+	godep restore > ${godepRestoreDir}/godep-restore.log 2>&1 || true
```
Comment (sttts) [inline hack/update-staging-godeps.sh]: can we do better than hiding and ignoring all errors?

Comment (sttts) [inline hack/update-staging-godeps.sh]: At least, hide those "no repository" errors and fail if there is anything else.

Comment (sttts) [inline hack/update-staging-godeps.sh]: I removed the `2>&1 || true` and it worked fine. In which case does one get errors?

**Comment by sttts** [comment]
Tested on Mac 👍

File: hack/update-staging-godeps.sh (hunk 16 above)
Comment (deads2k) [inline hack/update-staging-godeps.sh]: updated,

File: hack/update-staging-godeps.sh (hunk 15 above)
Comment (deads2k) [inline hack/update-staging-godeps.sh]: done

**Comment by deads2k** [comment]
comments addressed and squashed in.

**Comment by deads2k** [comment]
@k8s-bot unit test this

**Comment by deads2k** [comment]
@k8s-bot verify test this

**Comment by deads2k** [comment]
approved based on https://github.com/kubernetes/kubernetes/issues/40363

**Comment by sttts** [comment]
Conflicting files
vendor/BUILD 😱

**Comment by deads2k** [comment]
@k8s-bot kubemark e2e test this

**Comment by deads2k** [comment]
@k8s-bot unit test this

**Comment by deads2k** [comment]
@k8s-bot unit test this

**Comment by deads2k** [comment]
@k8s-bot verify test this

**Comment by deads2k** [comment]
This is a verify script.  Every time a dep changes, this has to be rebased.  Bumping priority

**Comment by deads2k** [comment]
```
go test -v k8s.io/kubernetes/test/integration/quota -run TestQuotaLimitedResourceDenial$
quota_test.go:333: unexpected error: insufficient quota to consume: pods
```

**Comment by deads2k** [comment]
@k8s-bot unit test this

**Comment by deads2k** [comment]
@k8s-bot non-cri e2e test this

**Comment by deads2k** [comment]
@k8s-bot gci gce e2e test this

**Comment by deads2k** [comment]
@k8s-bot unit test this
@k8s-bot gci gce e2e test this

**Comment by gmarek** [comment]
@k8s-bot gci gce test this

**Comment by gmarek** [comment]
@k8s-bot gce etcd3 e2e test this

**Comment by deads2k** [comment]
@k8s-bot verify test this

**Comment by gmarek** [comment]
State of submit queue makes me sad...

**Comment by gmarek** [comment]
I hate you
@k8s-bot cvm gce e2e test this

**Comment by deads2k** [comment]
@k8s-bot cvm gce e2e test this: issue #41891

**Comment by janetkuo** [comment]
Staging godeps takes a long time to update :(
Verification failed on my PR, so I updated it with the script. However, the diff looks unnecessary (only adds one more char to the comment)?https://github.com/kubernetes/kubernetes/pull/41851/commits/a8506813dcf88d957bcec7cf2d9905b366332b1f

**Comment by sttts** [comment]
@janetkuo David mentioned something that the godep version makes the difference. The script looks like that it pins it to v74. Can you verify (bash -x) that it actually succeeds with that?

**Pull request merged**